src/pstore/pstore.c

```diff
@@ -0,0 +1,399 @@
+                 * previous record id, then start a new dmesg.txt file */
+                if (!streq(current_id, id)) {
+                        /* Encountered a new dmesg group, close out old one, open new one */
+                        if (dmesg) {
+                                (void) write_dmesg(dmesg, dmesg_size, id);
+                                free(dmesg);
+                                dmesg = NULL;
+                                dmesg_size = 0;
+                        }
+
+                        /* Swap record id pointers */
+                        id[0] = '\0'; /* invalidate previous record id */
+                        id = (id == id1) ? id2 : id1; /* make current id */
+                }
+
+                /* Reconstruction of dmesg is done as a useful courtesy, do not log errors */
+                dmesg2 = realloc(dmesg, dmesg_size + strlen(pe->dirent.d_name) + strlen(":\n") + 1);
+                if (dmesg2) {
+                        dmesg = dmesg2;
+                        dmesg_size += sprintf(&dmesg[dmesg_size], "%s:\n", pe->dirent.d_name);
+                        if (pe->content) {
+                                dmesg2 = realloc(dmesg, dmesg_size + pe->content_size + 1);
+                                if (dmesg2) {
+                                        dmesg = dmesg2;
+                                        memcpy(&dmesg[dmesg_size], pe->content, pe->content_size);
+                                        dmesg_size += pe->content_size;
+                                }
+                        }
+                }
+        }
+        if (dmesg) {
+                (void) write_dmesg(dmesg, dmesg_size, id);
+                free(dmesg);
+                dmesg = NULL;
+                dmesg_size = 0;
+        }
+}
+
+static int list_files(PStoreList *list, const char *sourcepath) {
+        DIR *dirp;
+        struct dirent *de;
+        PStoreEntry *pe;
+        int r = 0;
+
+        dirp = opendir(sourcepath);
+        if (!dirp)
+                return log_error_errno(errno, "Failed to opendir %s: %m", sourcepath);
+
+        FOREACH_DIRENT(de, dirp, return log_error_errno(errno, "Failed to iterate through %s: %m", sourcepath)) {
+                _cleanup_free_ char *ifd_path;
+
+                ifd_path = path_join(sourcepath, de->d_name);
+                if (!ifd_path)
+                        return log_oom();
+
+                _cleanup_free_ char *buf = NULL;
+                size_t buf_size;
+
+                /* Now read contents of pstore file */
+                r = read_full_file(ifd_path, &buf, &buf_size);
+                if (r < 0) {
+                        log_warning_errno(r, "Failed to read file %s: %m", ifd_path);
+                        continue;
+                }
+
+                if (!GREEDY_REALLOC(list->entries, list->n_entries_allocated, list->n_entries + 1))
+                        return log_oom();
+
+                pe = &list->entries[list->n_entries++];
```

Comment:
```c
list->entries[list->n_entries++] = (PStoreEntry) {
        .dirent = *de,
…
```

i.e. C99 structured initialization makes this so much nicer!

Reply:
done